feat: monorepo

vanpho93 opened the pull request:
Resolves: #6389
Impact: minor
Type: feature

## Issue

Too many repos.

## Solution

Using changeset and npm workplace to convert things into monorepo

If you're solving a UIX related issue, please attach screen-caps or gifs showing how your solution differs from the issue.

aldeed committed:
feat: initial project

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

spencern committed:
chore: add DCO instructions and LICENSE file

Signed-off-by: Spencer Norman <spencern@gmail.com>

aldeed committed:
feat: add setToExpectedIfMissing option

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

aldeed committed:
chore: rename to db-version-check

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

kieckhafer committed:
Merge pull request #1 from reactioncommerce/feat-aldeed-set-version

Add setToExpectedIfMissing and rename package

aldeed committed:
chore: CI publish

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

aldeed committed:
chore: add semantic release branch config

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

aldeed committed:
docs: document setToExpectedIfMissing

Signed-off-by: Eric Dobbertin <eric@dairystatedesigns.com>

vanpho93 committed:
feat: add DS_Store to .gitignore

vanpho93 committed:
feat: move root reaction to apps directory

vanpho93 committed:
feat: add changesets

vanpho93 committed:
feat: add github action workflows

vanpho93 committed:
feat: add package json to root and update reaction package json

vanpho93 committed:
feat: update package list

vanpho93 committed:
feat: upgrade logger and random packages

vanpho93 committed:
feat: update eslintrc file

vanpho93 committed:
feat: update package.json file

vanpho93 committed:
feat: update file-collections

vanpho93 committed:
fix: fix eslint

vanpho93 committed:
fix: fix post install

vanpho93 committed:
feat: update package json for apps/reaction

vanpho93 committed:
fix: lint gql

vanpho93 committed:
refactor

vanpho93 committed:
feat: readd dockerfile

vanpho93 committed:
Upgrade striple

vanpho93 committed:
feat: add some scripts

vanpho93 committed:
feat: remove test script from payments example

vanpho93 committed:
feat: add prepublishOnly script for logger and random

vanpho93 committed:
feat: rename package org

vanpho93 committed:
feat: remove dotenv from apps/reaction

vanpho93 committed:
feat: udd dist to root .gitignore

vanpho93 committed:
feat: use npm 8.1.0

vanpho93 committed:
chore: change repo inside changelog

Signed-off-by: Brian Nguyen <vanpho02@gmail.com>

aldeed commented on .changeset/config.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "https://unpkg.com/@changesets/config@2.1.1/schema.json",
+  "changelog": ["@changesets/changelog-github", { "repo": "vanpho93/reaction-monorepo" }],
```
Need to change to correct `repo`

aldeed commented on .circleci/config.yml:
```diff
@@ -53,7 +59,9 @@ jobs:
-          command: npm run lint
+          command: |
+            npm run build:packages
+            npm run lint
```
Is building necessary to run eslint and tests? Usually it is not.

vanpho93 replied:
![image](https://user-images.githubusercontent.com/17959025/186926500-c818d144-4324-41d2-988e-2f260ec016a8.png)
Without `build:packages` there will be an error. Maybe there's a way around to fix this.

aldeed commented on .circleci/config.yml:
```diff
@@ -125,10 +149,13 @@ jobs:
             - reaction-v6-node-modules-{{ .Branch }}
+      - run:
+          name: Authenticate with registry
+          command: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > .npmrc
```
This will overwrite the contents of `.npmrc`. You might want to do `> ~/.npmrc` to avoid losing the contents of the repo file, and/or do `>>` instead of `>` to append to the existing file contents.

aldeed commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,28 @@
+        with:
+          publish: npm run release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # Ensure to have this set up under GitHub secrets
```
You don't set `GITHUB_TOKEN` in secrets. Instead, you add `permissions` object on this workflow job, and then `GITHUB_TOKEN` will be automatically available with those permissions. Docs: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

vanpho93 replied:
We don't manually set GITHUB_TOKEN in secrets, this is just get the value only.

aldeed replied:
I was referring to the comment, which can be removed. And also I didn't see a `permissions` section in the yaml at the time, but maybe the default permissions are fine.

aldeed commented on package.json:
```diff
@@ -1,51 +1,21 @@
-  "name": "reaction-api",
-  "version": "4.2.0",
+  "name": "reaction",
+  "version": "1.0.0",
```
Need to decide what version this should be. Or omit the version property.

aldeed commented on package.json:
```diff
@@ -1,51 +1,21 @@
+  "scripts": {
+    "start:dev": "npm run start:dev -w=apps/reaction",
+    "start:meteor-blaze-app": "npm run start -w=apps/meteor-blaze-app",
+    "build:packages": "npm run build -w=packages/file-collections -w=packages/file-collections-sa-base -w=packages/file-collections-sa-gridfs -w=packages/logger -w=packages/random",
```
I think it would be better to do `npm run build --if-present --ws` so that it will build any new packages automatically.

aldeed commented on package.json:
```diff
@@ -1,51 +1,21 @@
+    "release": "changeset publish",
+    "lint": "eslint . --resolve-plugins-relative-to .",
+    "lint:docker": "npx --quiet --package @reactioncommerce/ci-scripts@1.6.2 lint-dockerfiles --workspace=apps/reaction",
+    "lint:gql": "npm run lint:gql -w=apps/reaction -w=packages/api-core -w=packages/api-plugin-surcharges",
```
I think it would be better to do `npm run lint:gql --if-present --ws` so that it will lint GraphQL for all packages that have it set up.

aldeed reviewed (changes requested):
Looks pretty good! I left some inline comments, and you can delete .circleci and .github folders from every package/app folder.

vanpho93 committed:
fix: address feedbacks

Signed-off-by: Brian Nguyen <vanpho02@gmail.com>

vanpho93 committed:
fix: update version fields

Signed-off-by: Brian Nguyen <vanpho02@gmail.com>

brent-hoover commented on apps/reaction/tests/integration/api/mutations/addTag/addTag.test.js:
```diff
@@ -154,14 +154,11 @@ describe("authorized user", () => {
   });
   test("cannot create tag with existing slug", async () => {
     const logLevel = Logger.level();
-    const duplicateTagSlug = Factory.Tag.makeOne({
```
Confused why this change is here?

vanpho93 replied:
Fixed by adding shopId. This change caused by an unique index changed from `slug: 1` into `slug: 1, shopId: 1`.

brent-hoover commented on apps/reaction/tests/integration/api/mutations/updateTag/updateTag.test.js:
```diff
@@ -182,7 +182,7 @@ describe("authorized user", () => {
     });
     Logger.level("FATAL");
-    expect.assertions(1);
+    // expect.assertions(1);
```
Why is this expect removed?

brent-hoover commented on packages/api-core/package.json:
```diff
@@ -50,7 +51,7 @@
   "dependencies": {
     "@apollo/federation": "^0.15.0",
     "@graphql-tools/merge": "^8.0.1",
-    "@reactioncommerce/api-utils": "^1.12.0",
```
Is this actually what we want here?

brent-hoover commented on packages/api-plugin-files/package.json:
```diff
@@ -21,24 +22,22 @@
     "@reactioncommerce/random": "~1.0.2",
     "@reactioncommerce/reaction-error": "~1.0.1",
-    "envalid": "~5.0.0",
+    "envalid": "^7.3.1",
```
We should be keeping the version of envalid consistent at 6.x.x

brent-hoover commented on packages/api-plugin-inventory-simple/src/index.js:
```diff
@@ -45,6 +45,7 @@ export default async function register(app) {
     policies,
-    queries
+    queries,
+    additionalFields: { test: 4 }
```
I think this is leftover test code?

brent-hoover commented on packages/api-plugin-pricing-simple/src/index.js:
```diff
@@ -45,6 +45,7 @@ export default async function register(app) {
       PriceRange
-    }
+    },
+    additionalFields: { test: 4 }
```
More test code?

brent-hoover commented on packages/api-utils/package.json:
```diff
@@ -70,8 +71,7 @@
     "graphql": "^15.0.0",
-    "jest": "^26.0.1",
-    "semantic-release": "^17.0.4"
+    "jest": "^26.0.1"
```
Are we keeping the version of this consistent?

vanpho93 replied:
There are many inconsistent dependencies, I am not sure if we could solve them all in this PR.

brent-hoover replied:
Yeah, but we're touching this one in this PR

brent-hoover replied:
Sorry, misread the diff. Disregard the comment. But we should do a push to make all common dependencies consistent.

vanpho93 replied:
I updated jest and eslint. i think we can do more later.

brent-hoover reviewed (commented):
Just have a few questions

vanpho93 committed:
fix: other concerns

vanpho93 committed:
fix: revert unwanted change in api-core package.json

vanpho93 committed:
feat: update reaction-error package

vanpho93 committed:
feat: remove jest and eslint from sub package

tedraykov commented:
The Circle CI run doesn't seem to be completely configured. We don't want to run linting and unit testing on every package on every commit, because it is redundant. Most CI/CD tools provide some sort of path filtering that run jobs only for the changed directories. 

Another concern I have is if we want to release a new docker image for `reaction`, we must be sure the integration tests that are running before the `docker-build-push` are using the newly published api-plugins.
Let's say we update the `orders` plugin and we run `npm run changeset` to generate the changesets. The changeset bot will generate a PR that bumps the `orders` plugin version and also update that version in the `reaction` api. When we merge this `release` PR, the circle ci and the github actions runs will start in parallel. A problem that might arise is when circle ci tries to build the `reaction` app that depends on the new `orders` plugin version, that might not be published by the github action job.

vanpho93 committed:
feat: update api-plugin-sample-data

vanpho93 committed:
feat: using jest 25

vanpho93 committed:
feat: update and commit new package-lock file

feat: remove package-lock.json for all packages

vanpho93 committed:
feat: update and commit new package-lock file

vanpho93 commented:
> The Circle CI run doesn't seem to be completely configured. We don't want to run linting and unit testing on every package on every commit, because it is redundant. Most CI/CD tools provide some sort of path filtering that run jobs only for the changed directories.

This's an improvement but we can do it later, there is a (tool)[https://turborepo.org/] that @aldeed suggested which could help.

> Another concern I have is if we want to release a new docker image for reaction, we must be sure the integration tests that are running before the docker-build-push are using the newly published api-plugins.

Integration tests run outside of docker.
There are still problem when a plugin upgrade trigger reaction app upgrade. In that case, `changeset` will update only package.json but not package-lock.json. I am trying to fix this.

@tedraykov @zenweasel

aldeed reviewed (changes requested):
I'm not sure about CircleCI, but for GitHub Actions, there is a good action plugin, `technote-space/get-diff-action`, that you can use to check folder diffs before running additional per-project checks.

```yaml
jobs:
  project-changed:
    runs-on: ubuntu-latest
    outputs:
      didChange: ${{ steps.projectDiff.outputs.count > 0 }}
    steps:
      - uses: actions/checkout@v3
      - name: Find changed files
        id: projectDiff
        uses: technote-space/get-diff-action@v6
        with:
          # List files and/or folders that should trigger remaining jobs when changed
          PATTERNS: |
            packages/package/**

  next-job:
    runs-on: ubuntu-latest
    # Skip if no project files have changed
    needs: project-changed
```

Note that it usually isn't correct to use the CI `paths` feature because that only checks the most recent commit, whereas you would want to run jobs if _any_ commit included in a PR has modified files in that folder.

My other concerns are:
- I don't know if shrinkwrap works with all of this. `api-plugin-authentication` is the only package that uses shrinkwrap. It's being used to pin a subdependency to an older version. You should remove the shrinkwrap file and instead use the newer [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) feature.
- To gain full advantages of the monorepo, you should migrate npm locks to the root package-lock and remove package-lock files from all packages.
- Did you test whether developing locally with multiple layers of package dependencies is working? For example, if running the app, which depends on `file-collections-sa-gridfs`, which in turn depends on `file-collections-sa-base`, can you edit `file-collections-sa-base` with hot reloading?

aldeed reviewed (changes requested):
A few more suggestions...

In the root `package.json`, add:

```json
"engines": {
    "node": ">=14.18.1",
    "npm": ">=7"
  },
  "engineStrict": true,
```

To support NPM v8.6+, run the following command and commit ([explanation](https://github.com/npm/cli/issues/4664#issuecomment-1086096726)):

```txt
npm config set legacy-peer-deps=true --location=project
```

(For this repo, there's a conflicting `graphql` version for which peer warnings were previously ignored. Eventually you could update packages to make everything happy and then migrate to strict peer deps.)

When I run `npm run start:dev`, there is a package error `Cannot find module 'json-bigint'`, which is coming from a dependency of `tus-node-server`. You can fix it by updating `tus-node-server` to `"~0.7.1"` in `packages/file-collections/package.json` and then re-run `npm install` in repo root. It seems like there are not any breaking changes with this update, but you should re-test file uploads to be sure. You may also need to update to a newer `tus-js-client`, which might have breaking changes.

In the new section of the readme, you are missing two steps for being able to run the app:
- Copy `apps/reaction/.env.example` to `apps/reaction/.env`.
- Start MongoDB. Maybe you can add a `docker-compose.yml` file in the repo root that just starts mongo, so they can run `docker-compose up -d` before `npm run start:dev`? It would be similar to the `docker-compose.yml` file in `apps/reaction`, but without the app service. You would also have to change the MONGO_URL to `mongodb://localhost:27017/reaction` in the `.env.example` file.

brent-hoover commented:
If we add a docker-compose with Mongo, we could use the one from the CLI and the CLI could grab it from here

vanpho93 committed:
feat: add docker-compose file and update dependencies for the file-collection

brent-hoover commented on README.md:
```diff
@@ -60,6 +60,44 @@ When make completes, three services will be running on localhost:
+## What you need
+* NPM >= 7 (support workspaces)
+* Docker & docker-compose
+* MongoDB
```
This is incorrect, you don't need Mongo installed if you are using the docker-compose and most people are better of not installing it locally so you can have different instances of it

vanpho93 committed:
fix: failing integration tests

vanpho93 committed:
Add 'packages/db-version-check/' from commit '8b3e0973fa88db8c20249d68964b4742f9c4b992'

git-subtree-dir: packages/db-version-check
git-subtree-mainline: 12013cf8dc7893c4a05c4e28aacd95d1ea957ecb
git-subtree-split: 8b3e0973fa88db8c20249d68964b4742f9c4b992

vanpho93 committed:
fix: missing changes in db version check package.json

vanpho93 committed:
feat: move to pnpm

brent-hoover commented on .github/workflows/release.yml:
```diff
@@ -12,21 +12,30 @@ jobs:
-      - name: Install npm 8.1.0
-        run: npm i -g npm@8.1.0
+      - name: Install pnpm
+        run: npm i -g pnpm@latest
```
I think we had agreed to pin npm at a particular version? There has been problems with it pulling in a breaking version of npm

brent-hoover commented on packages/api-plugin-sample-data/package.json:
```diff
@@ -10,6 +10,7 @@
   "homepage": "https://github.com/reactioncommerce/reaction",
   "url": "https://github.com/reactioncommerce/reaction.git",
   "email": "hello-open-commerce@mailchimp.com",
+  "private": true,
```
I think this disables publishing to npm, which is something we do want?

brent-hoover commented on packages/api-core/src/importPluginsJSONFile.js:
```diff
@@ -41,7 +41,11 @@ export default async function importPluginsJSONFile(pluginsFile, transformPlugin
       throw new Error(`Plugin "${name}" is not set to a string`);
     } else if (/[a-zA-Z@]/.test(pluginPath[0])) {
-      ({ default: plugin } = await import(pluginPath));
+      try {
```
What's the reasoning for this?

vanpho93 committed:
feat: update Dockerfile

vanpho93 committed:
feat: new mongo image

tedraykov commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,41 @@
+      - name: Create Release Pull Request or Publish to npm
+        uses: changesets/action@v1
+        with:
+          publish: pnpm release
```
I'm skeptical this would work as intended. First of all there is no pnpm command called release, also there are no scripts called `release` in the plugins. When we merge this in trunk, are we sure what part of the repository will be published?

vanpho93 replied:
Actually release will be called on circleci on `pnpm publish -r`.

tedraykov commented on Dockerfile:
```diff
@@ -1,5 +1,36 @@
-FROM reactioncommerce/node-prod:14.18.1-v1
+# syntax=docker/dockerfile:1.4
+
+FROM node:14.18.1-alpine as deps
```
This docker file is better suited for the app folder itself, not In the root of the repo. If you want to build a container image with the root context but use the `apps/reaction/Dockerfile` you can specify this in the `docker build` command like so:
```
docker build -t ... -f ./apps/reaction/Dockerfile .
```
